src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2515,29 +2598,32 @@ private ImmutableArray<Diagnostic> GetSourceDeclarationDiagnostics(SyntaxTree sy
             }
 
             // NOTE: Concatenate the CLS diagnostics *after* filtering by tree/span, because they're already filtered.
-            ImmutableArray<Diagnostic> clsDiagnostics = GetClsComplianceDiagnostics(syntaxTree, filterSpanWithinTree, cancellationToken);
+            ImmutableBindingDiagnostic<AssemblySymbol> clsDiagnostics = GetClsComplianceDiagnostics(syntaxTree, filterSpanWithinTree, cancellationToken);
 
-            return result.AsImmutable().Concat(clsDiagnostics);
+            return new ImmutableBindingDiagnostic<AssemblySymbol>(result.AsImmutable().Concat(clsDiagnostics.Diagnostics), clsDiagnostics.Dependencies);
         }
 
-        private ImmutableArray<Diagnostic> GetClsComplianceDiagnostics(SyntaxTree syntaxTree, TextSpan? filterSpanWithinTree, CancellationToken cancellationToken)
+        private ImmutableBindingDiagnostic<AssemblySymbol> GetClsComplianceDiagnostics(SyntaxTree syntaxTree, TextSpan? filterSpanWithinTree, CancellationToken cancellationToken)
         {
             if (syntaxTree != null)
             {
-                var builder = DiagnosticBag.GetInstance();
+                var builder = BindingDiagnosticBag.GetInstance();
                 ClsComplianceChecker.CheckCompliance(this, builder, cancellationToken, syntaxTree, filterSpanWithinTree);
                 return builder.ToReadOnlyAndFree();
             }
 
-            if (_lazyClsComplianceDiagnostics.IsDefault)
+            if (_lazyClsComplianceDiagnostics.IsDefault || _lazyClsComplianceDependencies.IsDefault)
             {
-                var builder = DiagnosticBag.GetInstance();
+                var builder = BindingDiagnosticBag.GetInstance();
                 ClsComplianceChecker.CheckCompliance(this, builder, cancellationToken);
-                ImmutableInterlocked.InterlockedInitialize(ref _lazyClsComplianceDiagnostics, builder.ToReadOnlyAndFree());
+                var result = builder.ToReadOnlyAndFree();
+                ImmutableInterlocked.InterlockedInitialize(ref _lazyClsComplianceDependencies, result.Dependencies);
```

Comment:
With two interlocked exchanges here, this is no longer threadsafe. Please use an alternate pattern, such as making the field a `BindingDiagnosticsBag` instead of being two separate fields.

Reply:
Or, if you think this is safe, please leave a detailed comment explaining the reasoning in code.

---
In reply to: [361727949](https://github.com/dotnet/roslyn/pull/40560#discussion_r361727949) [](ancestors = 361727949)

Reply:
> With two interlocked exchanges here, this is no longer threadsafe. 

Please explain why it is not thread safe.

---
In reply to: [361728003](https://github.com/dotnet/roslyn/pull/40560#discussion_r361728003) [](ancestors = 361728003,361727949)

Reply:
> Please explain why it is not thread safe.

Because Thread 1 can initialize the first field, and thread 2 can initialize the second field. The way that this code is structured implies that both fields should be initialized by the same originator, which is no longer true. If you think that this is fine, please add add an assert to verify.

---
In reply to: [361731374](https://github.com/dotnet/roslyn/pull/40560#discussion_r361731374) [](ancestors = 361731374,361728003,361727949)

Reply:
> Because Thread 1 can initialize the first field, and thread 2 can initialize the second field. The way that this code is structured implies that both fields should be initialized by the same originator, which is no longer true. If you think that this is fine, please add add an assert to verify.

I will not be adding any comments or asserts here. It is obvious that the values are calculated the same way and the result is going to be logically the same regardless of the thread. 

---
In reply to: [361731858](https://github.com/dotnet/roslyn/pull/40560#discussion_r361731858) [](ancestors = 361731858,361731374,361728003,361727949)